### PR TITLE
Signup flow

### PR DESCRIPTION
**Pull request by nadasuhailAyesh12**

### CODE REVIEW

**Review by AhmedSafi97 — DISMISSED** [review]

**Comment by AhmedSafi97** [comment]
@nadasuhailAyesh12 resolve conflict please

**Review by fadezak100 — CHANGES_REQUESTED** [review]
I'll write the requested changes after the session.

File: src/interfaces/UserInterface.ts [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
 
 export interface UserInterface {
-  id: string
+  id?: string
   name: string
   email: string
   password: string
-  slides: SlideInterface[]
+  slides?: SlideInterface[]
+}
```
Comment (fadezak100) [inline src/interfaces/UserInterface.ts:10]: The `id` is not optional here.

File: src/interfaces/UserInterface.ts [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
+export interface UserOptions {
+  id?: number
+  email?: string
+  name?: string
+}
+export interface UserRequest extends Request {
+  body: UserInterface
```
Comment (fadezak100) [inline src/interfaces/UserInterface.ts:17]: As the `id` is auto generated, and you don't need it in the body request, you can use the Omit function to take it out from the UserInterface and still use the same interface with the request body.
``` Omit< UserInterface,  'id'>```

File: package.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "ca-projects-starter",
+  "name": "presently",
   "version": "1.0.0",
   "description": "",
   "main": "index.js",
```
Comment (fadezak100) [inline package.json:5]: Please delete the ```--detectOpenHandles``` from ```test``` script.

File: src/controllers/UserController.ts [hunk 4]
```diff
@@ -1 +1,29 @@
-export {}
+import { Response } from 'express'
+import UserService from '../services/UserService'
+import { signupvalidationschema } from '../validation/userValidation'
+import { validate } from '../validation/validator'
+import { UserRequest } from 'interfaces/UserInterface'
```
Comment (fadezak100) [inline src/controllers/UserController.ts:5]: Add the backspace by changing it to ```../interfaces/UserInterface```

File: src/controllers/UserController.ts [hunk 5]
```diff
@@ -1 +1,29 @@
...
+    })
+    if (!validationResult.isValid)
+      throw new GenericError(validationResult.error)
+
+    await UserService.createUser({ name, email, password })
+    res.status(201).json({ message: 'signup sucessfully' })
+  } catch (error: unknown) {
+    const exception = error as Error
+
+    if (exception.name !== 'GenericError') throw exception
```
Comment (fadezak100) [inline src/controllers/UserController.ts:24]: Just return the```return next(exception)```.  Make sure to import ```nextFunction``` in the controllers signatures.

File: src/services/UserService.ts [hunk 6]
```diff
@@ -1 +1,20 @@
-export {}
+import AuthHelper from '../helpers/AuthHelper'
+import { UserInterface } from '../interfaces/UserInterface'
+import UserRepository from '../repositories/UserRepository'
+import GenericError from '../helpers/GenericError'
+
+const createUser = async ({ name, email, password }: UserInterface) => {
```
Comment (fadezak100) [inline src/services/UserService.ts:6]: If you want to use the same interface, make sure to omit the `id` from the interface
```
const createUser = async ({ name, email, password }: Omit<UserInterface, id>) => 
```
Or you can create a new interface.

File: src/validation/userValidation.ts [hunk 7]
```diff
@@ -1 +1,7 @@
-export {}
+import * as yup from 'yup'
+
+export const signupvalidationschema = yup.object().shape({
+  name: yup.string().required(),
+  email: yup.string().email().required(),
+  password: yup.string().required().min(6).max(26)
```
Comment (fadezak100) [inline src/validation/userValidation.ts:6]: Add the password regex. You can copy it from Zayan's.

File: src/validation/validator.ts [hunk 8]
```diff
@@ -1 +1,23 @@
-export {}
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import yup from 'yup'
+
+export const validate = async ({
```
Comment (fadezak100) [inline src/validation/validator.ts:4]: change the export name to `validator`

**Review by fadezak100 — CHANGES_REQUESTED** [review]
Just small changes. After, fixing them make sure to pull the main and resolve conflict.

File: src/validation/validator.ts (hunk 8 above)
Comment (nadasuhailAyesh12) [reply src/validation/validator.ts:4]: Thank you, very great review I will do it as soon as possible

**Review by fadezak100 — APPROVED** [review]
LGTM. Pull the main branch and fix conflicts then we will be ready to merge.

**Review by AhmedSafi97 — APPROVED** [review]

**Pull request merged**